은행 창구 관리 앱  [STEP 1] Kyle, Effie

hyeffie opened the pull request:
두두(@FirstDo) 안녕하세요. 늦은 PR로 돌아온 에피(@hyeffie), 카일(@Changhyun-Kyle)입니다!

이번 스텝에서는 Queue 자료구조를 Linked List를 사용해 구현하고, 각 단위의 인터페이스를 테스트하는 경험을 했습니다. 자료 구조의 동작을 

## 요구사항 구현

### 자료 구조

- `Linked List`, `Node`
    - 일렬로 줄을 서는 은행 창구를 구현하기 위해 노드로 이루어진 단방향 연결 리스트를 구현했습니다.
    - `Queue`에서 사용하는 연산에 한 해 필요한 연산을 구현했습니다.
    - 유일한 상태인 `head`를 타입 외부에서 접근할 수 있되, 값을 변경할 수 없도록 `private(set)` 으로 구현했습니다.
- `Queue`
    - 내부적으로 `Linked List`를 상태로 갖는 큐를 구현했습니다.
    - `linked list`의 `head` 속성과 마찬가지로 외부에서 변경할 수 없도록 `private(set)`으로 구현했습니다.
    

### 단위 테스트

- `Queue`와 `Linked List`의 인터페이스에 대한 테스트를 구현했습니다.
- 각각의 연산의 정확한 역할에 집중해 테스트 메서드를 구현하려 노력했습니다.

---

## 고민한 점

### 특정 인터페이스를 검증하기 위해 같은 유닛의 다른 인터페이스를 호출해야 하는 문제

`Queue`의 각 인터페이스에 대한 테스트 메서드를 작성하다보니, 어떤 인터페이스를 테스트할 때 다른 인터페이스를 호출해야 하는 상황이 있었습니다. `clear`를 테스트하기 위해 `peek`을 사용해야 하거나, `enqueue`를 테스트 하기 위해 `dequeue` 를 사용해야 하는 상황이 그랬습니다.

```swift
// MARK: - Clear
    
func test_비어있지_않으면_clear했을때_LinkedList의_head가_nil이다() {
	// given
	setSUTWithOneElement("1")
        
	// when
	self.sut.clear()
        
	// then
	let result = self.sut.peek()
	XCTAssertNil(result)
}

// MARK: - Enqueue
    
func test_큐에_요소가_하나있을때_요소를_추가하면_요소가_rear에_추가된다() {
	// given
	let firstNode = Node(value: "old")
	self.sut = Queue(linkedList: .init(head: firstNode))

	// when
	let newValue = "new"
	self.sut.enqueue(newValue)

	// then
	var lastValue: String?
	while self.sut.peek() != nil {
		lastValue = self.sut.dequeue()
	}
	XCTAssertEqual(lastValue, newValue)
}
```

`Queue`에서 일어나는 연산에는 각각의 특징이 있습니다.

- `isEmpty`, `Peek`
    - 연산 프로퍼티(mutating이 아닌 인터페이스)
    - 테스트에서 다른 인터페이스를 사용할 호출할 필요 ❌
- `clear`, `enqueue`, `dequeue`
    - `mutating` 인터페이스
    - 테스트가 다른 인터페이스에 의존적

처음에는 각 테스트가 어떤 인터페이스에 의존하고 있는지를 파악해서 순서를 결정하고, 의존해야 하는 인터페이스를 순서상 먼저 작성하고, 이 인터페이스의 테스트가 성공하면 무결성이 입증되었다는 가정 하에 다른 인터페이스 테스트에서 사용하도록 작성했습니다.

하지만 이런 방식은 각 테스트 메서드의 독립성을 보장하지 못했고, 테스트 대상인 인터페이스를 다른 인터페이스의 테스트에서 호출하기 위해서는 결과적으로 상태에 의존할 수 밖에 없었습니다.

상태에 접근해 원하는 테스트 로직을 만들기 위해서는 `Queue`의 `linkedList`, `LinkedList`의 `head`와 같은 속성에 접근해야 했는데, 이 속성은 객체를 구현하면서 은닉하기 위해 `private`으로 접근을 제한한 상황이었습니다. 

```swift
struct LinkedList<Value> {
    private var head: Node<Value>?

    init(head: Node<Value>? = nil) {
        self.head = head
    }
}
struct Queue<Value> {
    private var linkedList: LinkedList<Value>
}
```

주요한 상태를 외부로 노출할 수 없다면 테스트를 위해 다른 인터페이스를 사용하는 방법도 있었지만, 객체의 관점에서 테스트만을 위한 `mutating`한 인터페이스를 만드는 방향은 객체에 대한 부작용을 발생시키고, `testable`한 코드가 아니기에 올바르지 않다고 생각했습니다.

객체의 상태에 접근하는 것이 부작용을 일으키지 않으려면(`mutating`하지 않으려면), `setter`에 대한 접근을 제한해야 한다고 판단해 각 속성에 `private(set)`을 추가했습니다.

```swift
struct LinkedList<Value> {
    private(set) var head: Node<Value>?
}

struct Queue<Value> {
    private(set) var linkedList: LinkedList<Value>
}
```

이를 통해, 아래와 같이 각 속성의 `getter`에 대한 접근 레벨을 낮추어, 테스트 모듈에서 각 객체의 속성에 접근해 테스트할 수 있게 되었습니다. 

```swift
// MARK: - Enqueue
//구현 전    
func test_큐에_요소가_하나있을때_요소를_추가하면_요소가_rear에_추가된다() {
	// given
	let firstNode = Node(value: "old")
	self.sut = Queue(linkedList: .init(head: firstNode))

	// when
	let newValue = "new"
	self.sut.enqueue(newValue)

	// then
	var lastValue: String?
	while self.sut.peek() != nil {
		lastValue = self.sut.dequeue()
	}
	XCTAssertEqual(lastValue, newValue)
}

//구현 후
func test_큐에_요소가_하나있을때_요소를_enqueue하면_요소가_rear에_추가된다() {
	// given
	setSUTWithOneElement("old")
        
	// when
	let newValue = "new"
	self.sut.enqueue(newValue)
        
	// then
	var lastNode = self.sut.linkedList.head
	while let nextNode = lastNode?.next {
		lastNode = nextNode
	}
	let result = lastNode?.value
	let expected = newValue
	XCTAssertEqual(result, expected)
}
```

---

## 질문

- 한 테스트 메서드 안에서 `assert`를 두 번 호출해도 될까요?
    
    ```swift
    func test_큐에_요소가_두개있을때_dequeue한_값은_front의_첫번째_요소이다() {
    	// given
    	let (firstValue, secondValue) = ("1", "2")
    	setSUTWithTwoElements(firstValue, secondValue)
            
    	// when
    	let result = self.sut.dequeue()
            
    	// then
    	XCTAssertEqual(result, firstValue)
    	XCTAssertNotEqual(result, secondValue)
    }
    ```
    

- 테스트 할 때 `initializer`를 인터페이스 호출해서 초기값을 설정해도 될까요? 이니셜라이저도 테스트 대상이 될 수 있다면 다른 인터페이스에 의존적인 상황이 될 것 같아서 고민됩니다.

- 현재 `Node`를 제외한 `LinkedList`와 `Queue` 타입이 구조체 입니다. `STEP 1` 명세서 단에서는 참조 혹은 상속을 필요로 하지 않기 때문에 구조체로 설계했습니다. 하지만, 구현을 위해 `mutating` 키워드 및 `initializer`를 활용해야했습니다.
    
    구조체의 불변성을 보장하지 못한다면, 복사의 이점을 제외하고는 구조체를 사용할 정확한 이유를 확보하지 못한 것 같은데, 이에 대해서 두두는 어떻게 생각하시는지 궁금합니다.

Changhyun-Kyle committed:
config: .gitignore -> .DS_Store 추가

hyeffie committed:
config: swift lint 관련 프로젝트 설정

- run script 추가
- lint rule 추가

hyeffie committed:
config: 프로젝트 권장 설정 업데이트

hyeffie committed:
feat: Node, LinkedList, Queue 구현

hyeffie committed:
chore: 구현 수정

hyeffie committed:
test: 테스트 추가

hyeffie committed:
test: 새 테스트 메서드 구현

- 은닉화

hyeffie committed:
refine: 내부 로직 개선 및 불필요한 인터페이스 제거

hyeffie committed:
Merge branch 'effie/step1/linked-list' into step1/linked-list

Changhyun-Kyle committed:
chore: 역할과 책임에 따른 객체 분리 및 파일 분리

Changhyun-Kyle committed:
chore: target membership 변경

Changhyun-Kyle committed:
refine: 접근제어자 변경

Changhyun-Kyle committed:
test: 유닛테스트 로직 수정

hyeffie committed:
test: LinkedList 테스트 생성

hyeffie committed:
test: unit 구현 변경에 대응한 테스트 작성

hyeffie committed:
test: 테스트 메서드 리팩토링

hyeffie committed:
Merge branch 'effie/step1/test' into step1/linked-list

Changhyun-Kyle committed:
chore: Mark 주석 추가 외

Changhyun-Kyle committed:
chore: Node 클래스 final 키워드 추가

Changhyun-Kyle committed:
chore: Node 클래스 final 키워드 추가

Changhyun-Kyle committed:
chore: 유닛테스트 파일 이름 변경 및 파일 순서 변경

hyeffie committed:
config: 프로젝트 signing - team None으로 설정

hyeffie committed:
Merge branch 'step1/linked-list' into develop

FirstDo commented on BankManagerConsoleApp/.swiftlint.yml:
```diff
@@ -0,0 +1,31 @@
+disabled_rules:
+- force_cast
+- trailing_whitespace
+- trailing_comma
+- class_delegate_protocol
+
+force_unwrapping:
+  severity: error
+
+function_body_length:
+- 15 # warning
+- 30 # error
+
+type_body_length:
+- 200 # warning
+- 400 # error
+
+line_length:
+- 150 # warning
+- 200 # error
+
+nesting:
+  type_level:
+    - 2
+
+opt_in_rules:
+- force_unwrapping
+- indentation_width
+- sorted_imports
+- multiline_arguments
+- multiline_parameters_brackets
```
swiftlinit를 사용하셨네요!
이런 툴들을 사용하면 컨벤션을 강제로 맞춰줄 수 있어서 큰 도움이 되죠!
swiftformat 이라는 툴도 있는데, swiftlinit와는 또다른 성격이라서 둘다 알아두심 도움이 될것 같습니다 ㅎㅎ

FirstDo commented on BankManagerConsoleApp/BankManagerConsoleApp/LinkedList.swift:
```diff
@@ -0,0 +1,41 @@
+        self.head = head
+    }
+    
+    func getFirst() -> Value? {
+        return head?.value
+    }
```
연산프로퍼티를 쓸 때와, 매서드를 쓸 때 어떤 기준으로 선택하시나요?

`isEmpty`는 연산프로퍼티, `getFirst()`는 매서드인데 두분이서 어떤 기준으로 이렇게 결정하셨는지가 궁금하네요!
제 생각으로는 둘이 다를 이유가 없는것 같아요

hyeffie replied:
파라미터가 필요하지 않은지, 연산 중에 mutating 메서드를 호출하지 않는지 등이 기준이 될 수 있겠지만,
위의 경우에는 기준 없이 작성했던 것 같아요. 구현 당시에는 Queue의 peek 메서드를 구현하기 위해서 LinkedList에 getFirst를 추가했던 것 같은데, 이 과정에서 두 개의 일관성을 위해 선택했던 것 같습니다...!

FirstDo replied:
둘다 상관은 없지만, 일관성 있는 코드 작성을 위해서 자기만의 기준을 가지고 있는게 중요하다고 생각됩니다!
참고로 저 같은 경우 단순조회는 연산프로퍼티로 하는 편입니다..!

FirstDo commented on BankManagerConsoleApp/BankManagerConsoleApp/LinkedList.swift:
```diff
@@ -0,0 +1,41 @@
+        return head == nil
+    }
+    
+    mutating func add(value: Value) {
+        guard let head else {
+            self.head = Node(value: value)
+            return
+        }
+        var current: Node = head
+        while let next = current.next {
+            current = next
+        }
+        current.next = Node(value: value)
+    }
```
링크드리스트의 길이가 n일때
Big-O표기법으로 add 매서드의 시간복잡도는 어떻게 될까요?
add를 수행할때마다 그 시간복잡도를 가지는게 괜찮을까요??

hyeffie replied:
밑에서 주신 코멘트를 반영하기로 결정해서 tail 속성을 추가하게 되었어요. tail에 접근함에 따라 시간복잡도가 O(n)에서 O(1)로 개선되었습니다!

FirstDo commented on BankManagerConsoleApp/BankManagerConsoleApp/LinkedList.swift:
```diff
@@ -0,0 +1,41 @@
+        current.next = Node(value: value)
+    }
+    
+    @discardableResult
+    mutating func removeFirst() -> Value? {
+        let result = head?.value
+        self.head = head?.next
+        return result
+    }
```
add 매서드와 달리 removeFirst는 상수시간에 수행이 가능하겠네요!
add도 그런식으로 만들어 볼 수 있을까요?

Changhyun-Kyle replied:
`tail` 프로퍼티를 추가하여 `add()` 메서드를 리팩토링 했습니다!
```swift
    mutating func add(value: Value) {
        let newNode = Node(value: value)
        guard let currentTail = self.tail else {
            self.head = newNode
            self.tail = self.head
            return
        }
        let previousTail = self.tail
        previousTail?.next = newNode
        self.tail = self.tail?.next
    }
```

FirstDo commented on BankManagerConsoleApp/BankManagerConsoleApp/Node.swift:
```diff
@@ -0,0 +1,16 @@
+//  Created by Effie on 1/22/24.
+//
+
+final class Node<Value> {
+    let value: Value
+    var next: Node?
+    
+    init(value: Value, next: Node? = nil) {
+        self.value = value
+        self.next = next
+    }
+}
```
질문에 대한 답변을 여기서 하겠습니다..!

```
현재 Node를 제외한 LinkedList와 Queue 타입이 구조체 입니다. 
STEP 1 명세서 단에서는 참조 혹은 상속을 필요로 하지 않기 때문에 구조체로 설계했습니다. 
하지만, 구현을 위해 mutating 키워드 및 initializer를 활용해야했습니다.

구조체의 불변성을 보장하지 못한다면, 복사의 이점을 제외하고는 구조체를 사용할 정확한 이유를 확보하지 못한 것 같은데, 이에 대해서 두두는 어떻게 생각하시는지 궁금합니다.
```

우선 제가 질문을 이해한 바로는 mutating때문에 키워드 때문에 구조체의 불변성이 보장이 안되는것 같다.
현재 스텝에서는 복사나 참조나 상관도 없고, 그렇게 때문에 굳이 구조채를 사용할 이유를 찾지 못한것같다.
제가 잘못이해한거라면 다시 dm으로 말씀해주세요

구조체 / 클래스의 장단점은 무엇이고, 어떤 기준으로 선택해야할까요?
두분이 생각하신점을 적어주시면 좋겠네요

저는 결국 `복사` or `참조`가 구조체와 클래스를 선택하는 가장 큰 이유가 되지 않을까 싶은데요,
우리가 만든 `LinkedList`를 한번 봅시다

우선 `Node`는 서로를 참조해야 하기때문에 클래스로 만들 수 밖에 없었죠?
그러면 LinkedList 라는 객체는 구조체인데 내부에 클래스 타입의 Node를 가지고 있게 될 겁니다.

아래 코드를 한번 볼까요?
```swift
var a = LinkedList<Int>(head: .init(value: 10))
var b = a

a.add(value: 10)

print(b.isEmpty)
```

a는 분명히 값 타입입니다
b는 a를 복사했고, 또 다른 값 타입이겠죠?

그 다음에 a에 10을 넣었습니다.
b.isEmpty 값이 어떻게 될까요?

분명히 독립적인 구조를 원해서 저런식으로 작성했을텐데, 
a,b 내부의 Node는 클래스 타입이므로 복사가 아니라 참조가 됩니다.

원했던 동작이 아니죠?
두분이 말씀하신 복사의 이점이 없어져 버렸네요.

관련된 WWDC가 있는데 지금보시기엔 약간 어려울것같아요.
추후에 좀 더 깊게 공부가 필요할때 보시면 좋을것 같아요.
https://developer.apple.com/videos/play/wwdc2016/416/

이런경우에는 저라면 final class를 쓸 것 같아요

hyeffie replied:
아 shallow copy로 동작한다는 사실을 잊고 있었어요. 공유해주신 영상은 학습 활동 때 학습 주제라서 카일이랑 저 둘 다 시청한 적이 있었는데, 영상의 내용대로 String처럼 내부에 참조를 저장하는 구조체의 경우 복사로 인해 여러 곳에 참조가 생기는 편이 더 나쁜 방향일 것 같습니다. 저희도 Queue와 LinkedList 모두 final class로 변경해보도록 하겠습니다!

Changhyun-Kyle replied:
> 구조체 / 클래스의 장단점은 무엇이고, 어떤 기준으로 선택해야할까요?
> 두분이 생각하신점을 적어주시면 좋겠네요

- 구조체: 값 타입으로 스택에 할당되어 static dispatch가 이루어져 적은 비용이 든다는 장점이 있습니다.
- 클래스: 참조 타입으로 스택에 메모리 주소가 저장되고 VTable을 통해 힙에 메모리가 할당되어 비용이 크지만, 상속 및 참조를 통해 다형성과 참조 타입의 이점을 가질 수 있습니다.
- 구조체와 클래스는 각각 Value Sementics와 Reference Sementics를 필요로 하는 상황에 적절히 선택하여 사용해야 합니다. 참조를 통해 객체가 공유돼야하는 상황에서는 클래스로 구현하고 참조를 사용하지 않고 값의 복사가 이뤄져야하는 상황에서는 구조체로 구현한다고 생각합니다.

FirstDo commented on BankManagerConsoleApp/BankManagerConsoleApp/Queue.swift:
```diff
@@ -0,0 +1,37 @@
+        return self.linkedList.removeFirst()
+    }
+    
+    mutating func clear() {
+        while self.isEmpty == false {
+            self.dequeue()
+        }
+    }
```
이부분도 시간복잡도 측면에서는 O(n)이 되겠네요
현재구조에서는 그냥 linkedList 자체를 다 날려버리고 다시 생성하는게 낮지 않을까 싶습니다.
linkedList를 클래스로 바꾼다면 또다른 방법을 찾아야겠지만요..

hyeffie replied:
새 인스턴스를 할당하는 방법을 사용할까 하다가 LinkedList의 head에 nil을 할당하는 방식으로 구현하고 관련 메서드와 테스트를 추가 수정했습니다!

FirstDo commented on BankManagerConsoleApp/BankManagerConsoleApp/Queue.swift:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+    
+    mutating func peek() -> Value? {
+        return self.linkedList.getFirst()
+    }
```
마찬가지로 peek()은 매서드고 isEmpty는 연산프로퍼티인 이유가 궁금합니다 :)

hyeffie replied:
위에서 설명했던 것과 같은 이유로 메서드가 아닌 연산 프로퍼티여도 문제는 없다고 생각합니다! 단지 메서드 이름에 따라서 그렇게 짓게 되었던 것 같아요.

FirstDo commented on BankManagerConsoleApp/BankManagerConsoleAppTest/QueueTest.swift:
```diff
@@ -0,0 +1,157 @@
+    
+    func test_큐가_비어있을때_peek을_하면_nil을_리턴한다() {
+        // given
+        setEmptySUT()
```
지금 모든 테스트 매서드에 setEmptySUT를 호출하고 있는데, 불필요해보입니다.

아래매서드의 역할이 뭐고, 언제 호출되는걸까요? 

```swift
    override func setUpWithError() throws {
        self.sut = Queue()
    }
```

hyeffie replied:
setUpWithError와 setUp 메서드는 각 테스트 메서드가 실행되기 전에 호출되어 테스트 이전에 테스트 대상(SUT)를 초기화하는 역할의 메서드라고 알고 있습니다!
처음에 템플릿에서 제공하는 메서드를 기본적으로 오버라이딩하고 있었고, 원래 각 테스트 메서드마다 사용할 sut를 설정해주는 코드를 given 섹션에서 구현했었는데요. 테스트 상황에 따라서 빈 Queue 등으로 초기화 해주는 setUp이 호출된다면 given이 필요하지 않아서, 이런 경우 setEmptySUT를 호출하지 않고 있었기 때문에 오버라이딩 구현을 유지했던 것 같습니다.
이런 경우, setUpWithError를 제거하는 게 맞을까요, 아니면 같은 동작을 하는 커스텀 메서드를 제거하고, given을 비워두는 게 맞을까요?

FirstDo replied:
누가 정답이라고 알려준적은 없지만,
기본제공하는 setUpWithError를 유지하고 given을 주석만 남기고 비워두는게 좀 더 적절하지 않나 싶습니다 ㅎㅎ

FirstDo commented on BankManagerConsoleApp/BankManagerConsoleAppTest/LinkedListTest.swift:
```diff
@@ -0,0 +1,163 @@
+}
+
+extension LinkedListTest {
+    private func setEmptySUT() {
+        self.sut = LinkedList<String>()
+    }
```
마찬가지로 불필요해 보입니다!

아래 매서드들은 언제 호출될까요?
```swift
    override func setUpWithError() throws {
        self.sut = LinkedList()
    }

    override func tearDownWithError() throws {
        self.sut = nil
    }
```

hyeffie replied:
setUpWithError와 setUp 메서드는 각 테스트 메서드가 실행되기 전에 호출되어 테스트 이전에 테스트 대상(SUT)를 초기화하는 역할의 메서드이고, 반대로 tearDownWithError는 다음 테스트 실행 사이클 이전에 호출되어 SUT와 관련된 작업을 정리해주는 역할이라고 알고 있습니다.

FirstDo commented on BankManagerConsoleApp/BankManagerConsoleAppTest/QueueTest.swift:
```diff
@@ -0,0 +1,157 @@
+        self.sut.enqueue(newValue)
+        
+        // then
+        var lastNode = self.sut.linkedList.head
+        while let nextNode = lastNode?.next {
+            lastNode = nextNode
+        }
+        let result = lastNode?.value
+        let expected = newValue
```
편의를 위해서 Queue의 앞을 front, 뒤를 rear라고 하겠습니다.
이부분도 다 이어지는거긴 한데, 결국 rear 조회를 위해서 접근제어자 수준을 변경하고 하신거잖아요?

위랑 이어지는 이야기인데, 자료구조인데 단순 조회가 O(1)이 아닌것부터 문제가 시작되는거같아요

queue.enqueue(10) 했으면
바로 queue.rear == 10 으로 비교할 수 있어야 하는게 아닌가 싶네요..!

hyeffie replied:
저희가 처음에 tail 이 linked list가 제공하는 기본 기능이라는 생각을 못하고 구현했어서 테스트에서 tail이 정말 필요한 순간마다 tail을 추가하는 것이 테스트만을 위한 인터페이스를 만드는 게 아닌가? 라고 생각했던 것 같아요. 말씀해주신 것처럼 값을 추가하는 기능등을 O(1)의 복잡도로 수행하기 위해서는 tail이 필수적으로 제공되어야 할 것 같더라고요.
tail과 Queue의 front와 rear 속성을 추가해서 낮은 비용으로 자료에 접근할 수 있도록 구현을 수정했습니다! 좋은 코멘트 감사합니다 ☺️

Changhyun-Kyle replied:
```swift
struct Queue<Value> {
    var front: Node<Value>? {
        return self.linkedList.head
    }
    
    var rear: Node<Value>? {
        return self.linkedList.tail
    }
}

struct LinkedList<Value> {
    private(set) var head: Node<Value>?
    
    private(set) var tail: Node<Value>?
}

// QueueTest
func test_큐에_요소가_하나있을때_요소를_enqueue하면_요소가_rear에_추가된다() {
      // given
      setSUTWithOneElement("old")
        
      // when
      let newValue = "new"
      self.sut.enqueue(newValue)
        
      // then
      let result = self.sut.rear?.value
      let expected = newValue
      XCTAssertEqual(result, expected)
}
```

FirstDo reviewed (changes requested):
안녕하세요 두분 고생하셨습니다!

LinkedList, Queue 두가지 자료구조를 구현해 주셨네요!
특히 PR문과 테스트 코드를 꼼꼼히 작성하셔서 보기도 편했습니다.

다만 자료구조는 시간복잡도가 상당히 중요한데, 
현재구조에서는 링크드리스트와 큐 모두 데이터를 추가할때나 맨끝 데이터를 조회할떄 (add, enqueue) 문제가 있어 보입니다.
더해서 Queue는 Double Stack으로도 간단하게 만들 수 있으니 그 방법도 알아두시면 좋을것 같아요

테스트코드에서도 불필요한 초기화가 반복적으로 이뤄지고 있는것 같아요..!

해당 부분들 개선 부탁드립나다 :)

## 질문에 대한 답변

### 1. 한 테스트 메서드 안에서 assert를 두 번 호출해도 될까요?

진짜 진짜 어쩔 수 없는 경우가 있는지는 모르겠지만, 그런 경우라면 그럴 수도 있을것 같아요
다만 예시로 들어주신 경우

`큐에_요소가_두개있을때_dequeue한_값은_front의_첫번째_요소이다`
를 판별하는건
`XCTAssertEqual(result, firstValue)` 로 충분하지 않을까요?
더이상 다른 판별문이 불필요해 보입니다.

### 2. 테스트 할 때 initializer를 인터페이스 호출해서 초기값을 설정해도 될까요? 이니셜라이저도 테스트 대상이 될 수 있다면 다른 인터페이스에 의존적인 상황이 될 것 같아서 고민됩니다.

우선  `initializer를 인터페이스 호출` 이 말이 무슨 뜻인지 제가 잘 이애하기 못했어요..!
설명 부탁드립니다 :)

이니셜라이저의 테스트는 이니셜라이저에서 복잡한 로직을 수행한다면, 객체의 생성 자체도 테스트 대상이 될 수는 있겠습니다만,
현재 프로젝트에 이니셜라이저들은 모두 단순 대입이기 때문에 고려하지 않으셔도 될것 같습니다
```swift
  init(head: Node<Value>? = nil) {
    self.head = head
  }
```

### 3. 현재 Node를 제외한 LinkedList와 Queue 타입이 구조체 입니다. STEP 1 명세서 단에서는 참조 혹은 상속을 필요로 하지 않기 때문에 구조체로 설계했습니다. 하지만, 구현을 위해 mutating 키워드 및 initializer를 활용해야했습니다.
구조체의 불변성을 보장하지 못한다면, 복사의 이점을 제외하고는 구조체를 사용할 정확한 이유를 확보하지 못한 것 같은데, 이에 대해서 두두는 어떻게 생각하시는지 궁금합니다.

해당 답변은 너무 길어지는것 같아서 코맨트에 달았습니다

나머지 자세한 것들은 코맨트에 남겨놨습니다!

Changhyun-Kyle committed:
refactor: LinkedList 및 Queue 속성 추가
- LinkedList: tail 속성 추가
- Queue: front 및 rear 속성 추가
- LinkedList: getFirst() 메서드 -> first 연산 프로퍼티로 변경
- LinkedListTest/Queue: first 연산 프로퍼티 적용

hyeffie committed:
refactor: 구현 변경에 따른 테스트 코드 수정

Changhyun-Kyle committed:
refactor: Queue 및  LinkedList 로직 수정refactor: Queue 및  LinkedList 로직 수정
- LinkedList & Queue: clear() 로직 추가 및 적용
- LinkedList: add() 메서드 불필요 로직 수정

Changhyun-Kyle committed:
refine: LinkedList Clear 테스트 코드 추가 및 테스트 함수명 수정

hyeffie committed:
remove: 하나의 테스트 메서드에 두 번 이상 assert 하는 코드 수정

hyeffie committed:
refactor: Queue, LinkedList 구조체 > final class로 변경

hyeffie committed:
Merge branch 'step1/linked-list' into develop

FirstDo commented on BankManagerConsoleApp/BankManagerConsoleApp/LinkedList.swift:
```diff
@@ -0,0 +1,46 @@
+        self.head = head
+    }
+    
+    func add(value: Value) {
+        let newNode = Node(value: value)
+        guard let currentTail = self.tail else {
+            self.head = newNode
+            self.tail = self.head
+            return
+        }
+        currentTail.next = newNode
+        self.tail = newNode
+    }
```
바뀐 링크드 리스트에 문제가 있습니다

Node에 deinit을 추가해서 디버깅 해봤습니다
```swift
  deinit {
    print(#function, value)
  }
```

```swift
let list = LinkedList<Int>(head: .init(value: 1))

list.add(value: 2)
list.add(value: 3)
list.add(value: 4)
list.add(value: 5)

print(list.first)

// 결과
// deinit 1
// Optional(2)
```

또다른 문제는 마지막 원소의 deinit이 호출되지 않습니다 (메모리가 남아있다는 말이겠죠?)

```swift
let list = LinkedList<Int>(head: .init(value: 1))

list.add(value: 2)
list.add(value: 3)
list.add(value: 4)
list.add(value: 5)

list.clear()

// 결과
// deinit 1
// deinit 2
// deinit 3
// deinit 4
```

힌트를 약간만 드리면
clear시 `마지막` Node만 해제가 안됬습니다. RC가 0이 아니라는 말이겠죠?

add시 `첫번째` Node가 바로 해제됬습니다.
그시점에 RC가 순간적으로 0이되었다는 말이겠죠?

hyeffie replied:
tail을 넣는다는 게 처리는 안 하고 있었더라구요 😅 initializer 랑 clear에 tail에 대한 처리를 추가해주었습니다!
```swift
final class LinkedList<Value> {
    private(set) var head: Node<Value>?
    
    private(set) var tail: Node<Value>?

    ...

    init(head: Node<Value>? = nil) {
        self.head = head
        var lastNode = self.head
        while let next = lastNode?.next {
            lastNode = next
        }
        self.tail = lastNode
    }

    ...    

    func clear() {
        self.head = nil
        self.tail = nil
    }
}
```

hyeffie committed:
fix: LinkedList init 및 clear 구현 수정

- init할 떄 head가 nil이 아닐 때도 tail을 설정해주지 않아서 add 시 head가 사라지던 문제
- clear 할 때 tail을 해제하지 않아서 tail인 노드만 사라지지 않던 문제
